packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.ts

```diff
@@ -30,8 +30,22 @@ export function maxByFirstDiffering<T>(
       return remainingValues[0];
     }
 
-    const max = Math.max(...remainingValues.map(fn));
-    remainingValues = remainingValues.filter((item) => fn(item) === max);
+    // Take a single pass through, accumulating all
+    // items with the single highest value.
+    let best: number = -Infinity;
+    const keep: T[] = [];
+    for (const item of remainingValues) {
+      const value = fn(item);
+      if (value < best) {
+        continue;
+      }
+      if (value > best) {
+        best = value;
+        keep.length = 0;
+      }
+      keep.push(item);
+    }
+    remainingValues = keep;
```

Comment:
I would be tempted to make this its own function for clarity, but I don't feel strongly. I guess there may be performance implications if this code is on the hot path

Reply:
Done. I'm still finding my feet for typescript performance, but I'm pretty confident that the JIT will figure out how to inline this. And empirically it is just as fast.